packages/use-presence/src/index.tsx

```diff
@@ -107,3 +109,57 @@ export default function usePresence(
     isExiting
   };
 }
+
+export type UseUniqueDataPresenceOptions<T> = {
+  /** Check if two instances of type T are equal, defaults to a === comparison */
+  equalityCheck?(a: T, b: T): boolean;
+  /** Check if an instance of type T is considered valid to render, defaults to Boolean function */
+  validationCheck?(a: T): boolean;
+} & UsePresenceOptions;
+
+function defaultEqualityCheck<T>(a: T, b: T) {
+  return a === b;
+}
+function defaultValidationCheck<T>(a: T) {
+  return Boolean(a);
+}
+
+export function useUniqueDataPresence<T>(
+  dataInput: T,
+  {
+    equalityCheck = defaultEqualityCheck,
+    validationCheck = defaultValidationCheck,
+    ...opts
+  }: UseUniqueDataPresenceOptions<T>
+) {
+  const [data, setData] = useState(dataInput);
+  const [shouldBeMounted, setShouldBeMounted] = useState(validationCheck(data));
+  const {isMounted, ...otherStates} = usePresence(shouldBeMounted, opts);
+  useEffect(() => {
+    if (!equalityCheck(data, dataInput)) {
+      if (isMounted) {
+        setShouldBeMounted(false);
+      } else if (validationCheck(dataInput)) {
+        setData(dataInput);
+        setShouldBeMounted(true);
+      }
+    } else if (!validationCheck(dataInput)) {
+      setShouldBeMounted(false);
+    } else if (validationCheck(data)) {
+      setShouldBeMounted(true);
+    }
+  }, [
+    dataInput,
+    data,
+    shouldBeMounted,
+    isMounted,
+    validationCheck,
+    equalityCheck
+  ]);
+
+  return {
+    ...otherStates,
+    isMounted: isMounted && validationCheck(data),
+    data
+  };
+}
```

Comment:
Can you add some tests for the new hook? The existing tests might help for inspiration. Also here moving to separate files per hook might be a good idea organization-wise.

Reply:
Done! Added to a separate test file `packages/use-presence/test/use-presence-switch.test.tsx` 👍